Pipenv -> poetry

haraschax opened the pull request:
- removes pipenv from openpilot
- add poetry, dependencies defined in `pyproject.toml`
  - `poetry install --no-root` so that "openpilot" isn't installed as a package itself...
- add xx packages to an optional [dependency group](https://python-poetry.org/docs/master/managing-dependencies/)
  - if `xx` dir is present, installs packages to system (`POETRY_VIRTUALENVS_CREATE=false`) and `--with xx`

haraschax committed:
Try poetry

haraschax committed:
Remove casadi for now

haraschax committed:
Update docker

haraschax committed:
Copy pipfiles

haraschax committed:
add casadi back

haraschax committed:
Too many slashes

haraschax committed:
New poetry api

haraschax committed:
Install system

haraschax committed:
Try again

haraschax committed:
No more pipenv

adeebshihadeh commented on update_requirements.sh:
```diff
@@ -3,7 +3,7 @@ set -e
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
 cd $DIR
-
+ls
```
remove this

adeebshihadeh commented on Dockerfile.openpilot_base:
```diff
@@ -12,19 +12,19 @@ ENV LANG en_US.UTF-8
 ENV PATH="$PYENV_ROOT/bin:$PYENV_ROOT/shims:$PATH"
 
-COPY Pipfile Pipfile.lock .python-version update_requirements.sh /tmp/
+COPY pyproject.toml poetry.lock .python-version update_requirements.sh /tmp/
```
is the cache still being remove properly here?

incognitojam replied:
I think the poetry cache is in `$HOME/.cache` so should be deleted

adeebshihadeh commented on update_requirements.sh:
```diff
@@ -26,22 +26,22 @@ eval "$(pyenv init --path)"
 
 if [ -d "./xx" ]; then
-  export PIPENV_SYSTEM=1
-  export PIPENV_PIPFILE=./xx/Pipfile
```
think this will need to be added back before merging?

JMPZ11 reviewed (changes requested):
Please also update tools/README.md - I'm still fairly green re: python, and if this changed without the readme, I would be lost... Thanks!

pd0wm commented:
Poetry 1.2 (with package groups) was just released: https://python-poetry.org/blog/announcing-poetry-1.2.0/

HaraldSchafer commented:
Cool! might try to revive this.

incognitojam committed:
new dependencies

incognitojam committed:
updates

incognitojam committed:
poetry 1.2.1, install dev dependencies

incognitojam committed:
keep install pipenv for xx for now?

incognitojam committed:
add pre-commit checks for poetry

incognitojam committed:
poetry lock is too slow

incognitojam committed:
update pip

incognitojam committed:
migrate to poetry groups

incognitojam committed:
update lockfile

incognitojam committed:
don't need to specify dev group unless it is made optional

incognitojam committed:
always install poetry

incognitojam committed:
set POETRY_VIRTUALENVS_CREATE instead, and use pipenv for xx

incognitojam committed:
use env for docs docker image

incognitojam commented on .pylintrc:
```diff
@@ -54,7 +54,7 @@ confidence=
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
-disable=C,R,W0613,W0511,W0212,W0201,W0106,W0603,W0621,W0703,W1201,W1203,E1136,W1514
+disable=C,R,W0613,W0511,W0212,W0201,W0106,W0603,W0621,W0703,W1201,W1203,E1136,W1514,E1124
```
Seems that pylint was updated, need to add this otherwise it complains about `video_link` and other keyword args in docs definitions

incognitojam commented on Dockerfile.openpilot_base:
```diff
@@ -12,19 +12,19 @@ ENV LANG en_US.UTF-8
 ENV LC_ALL en_US.UTF-8
 
-ENV PIPENV_SYSTEM=1
+ENV POETRY_VIRTUALENVS_CREATE=false
```
https://python-poetry.org/docs/configuration/#using-environment-variables

incognitojam committed:
alphabetical

incognitojam commented on update_requirements.sh:
```diff
@@ -40,25 +40,29 @@ fi
-  export PIPENV_SYSTEM=1
   export PIPENV_PIPFILE=./xx/Pipfile
-fi
+  pip install pipenv==2021.11.23
```
I think we still need to install pipenv for xx?

incognitojam commented on update_requirements.sh:
```diff
@@ -40,25 +40,29 @@ fi
+else
+  echo "PYTHONPATH=${PWD}" > .env
+  pip install poetry==1.2.1
+  poetry config virtualenvs.prefer-active-python true --local
```
since we use pyenv to manage the python installation we should set this
https://python-poetry.org/docs/managing-environments/

incognitojam replied:
it also creates a `poetry.toml` file - added to `.gitignore` since it can include custom user configuration

adeebshihadeh commented:
@incognitojam can you finish this up? goal is to get rid of `xx/blob/master/pipenv-merge.sh` in a clean way. if we can't do that, then let's just close this

incognitojam committed:
poetry 1.2.2

incognitojam committed:
add dev dependencies for typing

added in 4e310b80

incognitojam committed:
remove constraint

incognitojam committed:
fix PIPENV_SYSTEM

incognitojam committed:
remove constraint

incognitojam committed:
don't need this here

incognitojam committed:
bump

incognitojam committed:
bump pipenv

adds support for installing local pyprojects (can add openpilot as dependency of xx)

incognitojam committed:
README improvements

incognitojam commented:
@adeebshihadeh we could export the dependencies from the pyproject.toml to generate the xx Pipfile.. or I think adding openpilot as a dependency of xx would achieve the same thing?

incognitojam committed:
probably not necessary

adeebshihadeh commented on docs/docker/Dockerfile:
```diff
@@ -4,14 +4,16 @@ ENV PYTHONUNBUFFERED 1
+RUN pip install --no-cache-dir poetry==1.2.1 pip==22.2.2 && \
+    poetry install --no-ansi --no-interaction && \
+    pip uninstall -y poetry && \
+    rm -rf /root/.cache
```
hmm, is this even necessary? can we just remove instead? this should already be done in openpilot-base

incognitojam committed:
bump pip

adeebshihadeh commented on tools/README.md:
```diff
@@ -10,10 +10,9 @@ openpilot is developed and tested on **Ubuntu 20.04**, which is the primary deve
+git clone --recursive https://github.com/commaai/openpilot.git
 
-cd openpilot 
-git submodule update --init
```
can you keep this and add a comment like "openpilot uses git submodules, update them with:"? we should be better about including all the common workflow stuff here

incognitojam committed:
maybe not necessary?

incognitojam committed:
revert

incognitojam committed:
don't install openpilot itself into the virtual env

incognitojam committed:
remove PySide2 and shiboken2

reverts c6b749fb96b3b6332fd850e733140357f84d53a4

incognitojam committed:
remove Pipenv, add xx dependencies, sync system python

incognitojam committed:
add pipenv as xx dep

incognitojam committed:
semver package constraints, use old lockfile versions

incognitojam commented:
I've added semver version constraints to all of the `*` packages, using the version from the old `Pipfile.lock`, so they _shouldn't_ have breaking changes.. Also pinned the Python version to `~3.8`.

incognitojam committed:
fix casadi

incognitojam commented:
if CI passes this should be ready to review!

adeebshihadeh commented on update_requirements.sh:
```diff
@@ -40,25 +40,28 @@ fi
-  export PIPENV_PIPFILE=./xx/Pipfile
+  echo "WARNING: using xx dependency group, installing globally"
+  export POETRY_VIRTUALENVS_CREATE=false
+  POETRY_INSTALL_ARGS="--with xx --sync "
```
```suggestion
  POETRY_INSTALL_ARGS="--with xx --sync"
```

adeebshihadeh commented on update_requirements.sh:
```diff
@@ -40,25 +40,28 @@ fi
 
-echo "pip packages install..."
-pipenv sync --dev
-pipenv --clear
```
does poetry not have an equivalent of this?

incognitojam replied:
maybe we can do `pip cache purge` and `poetry cache clear --all PyPI`

adeebshihadeh replied:
this makes a big difference in CI, so I'd like that resolved before merging this

incognitojam replied:
I just tested it locally but it always prompted me to type `Y/n` even when I passed the `--no-interaction` option... supposedly fixed in Poetry 1.2.1 https://github.com/python-poetry/poetry/releases/tag/1.2.1

incognitojam replied:
we already do `rm -rf /root/.cache` in the Dockerfile which should clear this anyway?

adeebshihadeh replied:
what about installing with `--no-cache`?

incognitojam committed:
remove whitespace

Co-authored-by: Adeeb Shihadeh <adeebshihadeh@gmail.com>

incognitojam committed:
disable poetry cache

incognitojam committed:
cleanup

incognitojam committed:
prefer config file
